feat: G-API: Custom stream sources in Python (#27276)

ritamelo06 opened the pull request:
Implemented:

- A C++ proxy class PythonCustomStreamSource that implements the IStreamSource interface. This class acts as a bridge between G-API’s internal streaming engine and user-defined Python objects. Internally, it stores a reference to a Python object (PyObject*) and is responsible for: calling the Python object’s pull() method to retrieve the next frame, calling the descr_of() method to obtain the frame format description, acquiring and releasing the Python GIL as needed, converting the returned numpy.ndarray into cv::Mat, and handling any exceptions or conversion errors with proper diagnostics.

- A Python-facing factory function, cv.gapi.wip.make_py_src(), which takes a Python object as an argument and wraps it into a cv::Ptr<IStreamSource>. Internally, this function constructs a PythonCustomStreamSource instance and passes the Python object to it. This design allows Python users to define any class that implements two methods: pull() and descr_of(). No subclassing or special decorators are required on the Python side. The user simply needs to implement the expected interface.

This design enables any Python class with `pull()` and `descr_of()` to act as a custom stream source, no decorators or inheritance needed.

GSoC 2023 idea 15: [https://github.com/opencv/opencv/wiki/GSoC_2023#idea-g-api-implement-custom-stream-sources-in-python](url)

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [X] The PR is proposed to the proper branch
- [X] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [X] The feature is well documented and sample code can be built with the project CMake

asmorkalov commented:
Thanks a lot for the contribution! Could you add basic python test for the new binding?

ritamelo06 commented:
Thanks for the feedback! 
We’ve added a basic test for the new make_py_src binding in modules/gapi/misc/python/test/test_gapi_streaming.py. 
I’ve now also updated the checklist accordingly. Let me know if it needs any improvements,  we’ll be happy to make the necessary changes.

asmorkalov commented:
@AsyaPronina Could you take a look?

asmorkalov commented:
@TolyaTalamanov Could you take a look?

asmorkalov commented:
@AsyaPronina Could you take a look?
